### PR TITLE
feat: introduce major schema versioning

### LINKED ISSUE #360: Improve schema evolution process
### Description

Some schema migrations are extremely slow, possibly taking more than 24 hours, and will only get slower as we increase data volumes. This means we have to stop ingest and all database activity for long periods.

We need to distinguish the types of changes made to a schema:

 1. additive, non-breaking changes with no data migration (example: a new table, adding field comments, adding a view)
 2. additive, non-breaking changes with data migration (example: new materialized view, adding an index)
 3. destructive or breaking changes with no data migration (example: drop a table)
 4. destructive or breaking changes with data migration (example: change column type on populated table)

Cases 1 and 2 describe situations where the new schema still supports existing deployments of visor or querying clients such as grafana.

Cases 3 and 4 are situations where coordination effort is required to upgrade clients and visor instances.

Cases 2 and 4 are situations that could involve long periods of downtime as data is migrated.

We have discussed segregating schemas by task so they can be versioned independently but this only gives us control over the point at which we have to migrate, not the time taken. Besides, we usually want to run all tasks together so they are not very independent in reality.

We have also discussed using versioned tables but this has the disadvantage of being confusing for database users presented with multiple copies of tables. It can also lead to situations where the latest set of tables have different versions which makes it harder to keep external queries correct since every query will need to be inspected to ensure the correct versions are being used.

A simpler approach could be to have separate Postgres schemas with a versioned name containing a full set of the tables needed for visor. Updating external queries could be as simple as search and replace on the schema name, or using a user that defaults to the correct schema.

## Proposal

Introduce the notion of major schema versions and patches. Cases 2, 3 and 4 above are examples of changes that would indicate a major schema version change.  Only case 1 is considered a patch to an existing major schema.

Major schema versions should aggregate as many changes as possible with the aim of reducing the frequency of needing to halt indexing and client applications. A major schema change should be associated with at least a minor version increment for visor.

To minimise downtime and disruption, each major schema version should be deployed as a new postgres schema. All visor instances should be restarted pointing to the new schema to begin populating them with live data. Existing data is then migrated to the new tables in the background.

This means all query clients can continue using the previous schema but their views of the data will be frozen. When the data migration is complete the clients can be switched to the new schema and the old version may be removed.

The disadvantage of this approach is that we will need to maintain 2 copies of the full dataset for periods of time

Another disadvantage is that type 3 migrations, which would otherwise be quick to implement, get the overhead of a full data migration.

### PR DESCRIPTION
**Pull request by iand**
This is in preparation for a consolidation exercise on the database schema.  I've taken the proposal from https://github.com/filecoin-project/sentinel-visor/issues/360 and introduced structured versions. Each version now consistes of a `major` and a `patch`. The current schema has `major` 0. The `patch` is equal to the migration number, currently 28, so the current schema version is `0.28`. 

Migrations can be performed within the same major version, so 0.28 -> 0.29 is allowed but 0.28 -> 1.0 is not.

In Timescale the patch version is recorded using the `ID` column in the `gopg_migrations` table. The `major` version is `0` unless a table called `visor_version` exists with a single row containing the `major` version number. When we introduce schema `1.0` we will populate the `visor_version` table.

**Note:** migrations are now in versioned subfolders of `schemas`.

This PR also allows models to be aware of the schema version in use so they can adapt persistence logic when operating against an older schema version. This allows visor to run against older versions of the schema, avoiding the need to perform expensive migrations in lock step with deploying visor.

Note that this does not allow mixed versions to be used. Visor will detect the current schema version in the database and use that for all models.

Review notes: it's mostly mechanical changes wiring the version argument into the models. Some new tests cover the behaviour of selecting a version to persist. The database connect method uses a new constant to check the detected schema is not older than the oldest supported schema - which will be today's schema version.

The updated README in the schemas directory may also help orient yourself.

**Note:** I think we should now deprecate *down* migrations since we don't have any real use case for downgrading a schema except during testing

Fixes #360

### CODE REVIEW

**Comment by iand** [comment]
Test failures appear to be problems fetching vectors from ipfs

**Review by frrist — APPROVED** [review]
Awesome stuff -- :+1: to deprecating the down migration as well

File: storage/migrate.go [hunk 1]
```diff
@@ -21,50 +25,86 @@ func (d *Database) GetSchemaVersions(ctx context.Context) (int, int, error) {
...
 	return getSchemaVersions(ctx, db)
 }
 
 // getSchemaVersions returns the schema version in the database and the schema version defined by the available
 // migrations.
-func getSchemaVersions(ctx context.Context, db *pg.DB) (int, int, error) {
+func getSchemaVersions(ctx context.Context, db *pg.DB) (model.Version, model.Version, error) {
+	// Ensure the visor_version table exists
+	_, err := db.Exec(`
+		CREATE TABLE IF NOT EXISTS public.visor_version (
```
[the remark this answers is not in the document]
Comment (iand) [reply storage/migrate.go:39]: Will look at doing this in v1 schema #527

File: schemas/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# Schema Definitions
+
+This directory and its subdirectories holds the scripts for creating TimescaleDB database schemas compatible with the
+visor data model defined by the strucs in the `models` package.
+
+The data model and the corresponding database schemas are versioned using a `major` version number plus a `patc` number.
```
[the remark this answers is not in the document]
Comment (iand) [reply schemas/README.md:6]: done

File: schemas/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+For example `0.28`. Schemas with different major versions are not compatible with one another. Manual data migration 
+is required to transition from one major version to another.
+
+Patches are applied on top of the base schema for a major version and contain only additive, non-breaking changes with no data migration.
+This ensures pPatches are safe and can be applied by visor automatically. Some examples of additive, non-migration patches are adding a
```
[the remark this answers is not in the document]
Comment (iand) [reply schemas/README.md:11]: Done

File: schemas/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+**Do not modify existing migration files**
+
+1. Make the required changes to the Go models. 
+2. Ensure your test database is on the latest schema version (run `sentinel-visor migrate --latest`)
+3. Run the `TestSchemaIsCurrent` test in the storage package to compare the models to the current database schema. This will log the `CREATE TABLE` ddl for any altered tables.
+4. Create a new migration file using the next schema version as a prefix.
+5. Add all the statements required to migrate the schema from the previous version to an `up` variable.
+6. Add all the statements required to migrate the schema to the previous version to a `down` variable.
+7. Call `Patches.MustRegisterTx` with your `up` and `down` statements.
+8. Run the migration by running `sentinel-visor migrate --to <new-version>`
```
[the remark this answers is not in the document]
Comment (iand) [reply schemas/README.md:65]: Done

File: storage/csv.go [hunk 5]
```diff
@@ -24,10 +24,17 @@ var ErrMarshalUnsupportedType = errors.New("cannot marshal unsupported type")
 var (
 	// Cache of model schemas for csv storage
 	csvModelTablesMu sync.Mutex
-	csvModelTables   = map[string]table{}
+	csvModelTables   = map[nameWithVersion]table{}
 )
 
-func getCSVModelTable(v interface{}) table {
+type nameWithVersion struct {
+	name    string
```
[the remark this answers is not in the document]
Comment (iand) [reply storage/csv.go:31]: Done

File: storage/migrate.go [hunk 6]
```diff
@@ -75,17 +115,26 @@ func (d *Database) MigrateSchemaTo(ctx context.Context, target int) error {
...
 	}
 
 	if dbVersion == target {
 		return xerrors.Errorf("database schema is already at version %d", dbVersion)
 	}
 
-	if err := checkMigrationSequence(ctx, dbVersion, target); err != nil {
+	coll, err := collectionForVersion(target)
+	if err != nil {
+		return xerrors.Errorf("no schema definition corresponds to version %s", target)
```
[the remark this answers is not in the document]
Comment (iand) [reply storage/migrate.go:134]: Done

File: storage/migrate.go [hunk 7]
```diff
@@ -94,12 +143,17 @@ func (d *Database) MigrateSchemaTo(ctx context.Context, target int) error {
...
-	if dbVersion == 0 {
-		if _, err := db.Exec(`CREATE EXTENSION IF NOT EXISTS timescaledb WITH SCHEMA public`); err != nil {
-			return xerrors.Errorf("creating timescaledb extension: %w", err)
+	// Check if we need to create the base schema
+	if dbVersion.Patch == 0 {
+		log.Infof("creating base schema for major version %d", dbVersion.Major)
+
+		base, err := baseForVersion(dbVersion)
+		if err != nil {
+			return xerrors.Errorf("no base schema defined for version %s", dbVersion)
```
[the remark this answers is not in the document]
Comment (iand) [reply storage/migrate.go:152]: Done

**Pull request merged**